packages/server/src/common/command/command-stack.ts

```diff
@@ -13,62 +13,150 @@
+     * Removes the topmost (i.e. last executed) command from the stack and invokes {@link Command.undo}.
      */
     undo(): void;
 
     /**
-     * Moves the top of the stack up, redoing the new top command.
+     * Returns `true` if the top command on the stack can be undone.
+     */
+    canUndo(): boolean;
+
+    /**
+     * Re-adds the last undo command on top of the stack and invokes {@link Command.redo}
      */
     redo(): void;
 
     /**
-     * The command stack has currently unsaved changes.
+     * Returns `true` if there are  redoable commands in the stack.
      */
-    isDirty(): boolean;
+    canRedo(): boolean;
 
     /**
      * Called after a save has been successfully performed.
      */
     saveIsDone(): void;
+
+    /**
+     * Indicates wether the command stack has unsaved changes.
+     */
+    readonly isDirty: boolean;
+
+    /**
+     * Disposes all the commands in the stack.
+     */
+    flush(): void;
 }
 
 @injectable()
 export class DefaultCommandStack implements CommandStack {
+    @inject(Logger)
+    protected logger: Logger;
+
+    protected commands: Command[] = [];
+
+    /**
+     * The current position within the command list from which the next execute, undo, or redo, will be performed.
+     */
+    protected top = -1;
+
+    /**
+     * The current position within the command list when {@link DefaultCommandStack.saveIsDone} is called.
+     */
+    protected saveIndex = -1;
+
     execute(command: Command): void {
-        // no-op
+        try {
+            command.execute();
+        } catch (error) {
+            this.handleError(error);
+        }
+        // Clear the command list past the top (i.e. the old redo list)
+        this.commands = this.commands.slice(0, this.top + 1);
+        this.commands.push(command);
+        ++this.top;
+
+        // if the saveIndex points to the old redo list we can never reach a state again
+        // where save is not necessary. => ensure that 'isSaveNeeded' always returns true
+        if (this.saveIndex >= this.top) {
+            this.saveIndex = -2;
+        }
     }
 
     undo(): void {
-        // no-op
+        if (this.canUndo()) {
+            const command = this.commands[this.top--];
+            try {
+                command.undo();
+            } catch (error) {
+                this.handleError(error);
+            }
+        }
+    }
+
+    canUndo(): boolean {
+        return this.top !== -1 //
+            ? Command.canUndo(this.commands[this.top])
+            : false;
     }
 
     redo(): void {
-        // no-op
+        if (this.canRedo()) {
+            const command = this.commands[++this.top];
+            try {
+                command.redo();
+            } catch (error) {
+                this.handleError(error);
+
+                // Clear the command list past the top (i.e. the old redo list)
+                this.commands = this.commands.slice(0, this.top + 1);
```

Comment:
This is unreachable code as the `handleError` method will always throw an error. Also it seems to me that the handleError will always flush the whole command stack so there is no point in slicing anything.

Reply:
Good catch, no need to slice in the redo method.
In the execute method, however, we need the slicing because after successful execution of a new command we have to clear all "old" redo commands.

Reply:
Got it! :+1: